<a href="https://colab.research.google.com/github/jisu1013/kaggle_practice/blob/main/santander_product_recommendation_01_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
trn = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Santander Product Recommender/train_ver2.csv/train_ver2.csv')

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (5,8,11,15) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
trn.shape

(13647309, 48)

In [4]:
trn.head()

fecha_dato  ncodpers  ... ind_nom_pens_ult1 ind_recibo_ult1
0  2015-01-28   1375586  ...               0.0               0
1  2015-01-28   1050611  ...               0.0               0
2  2015-01-28   1050612  ...               0.0               0
3  2015-01-28   1050613  ...               0.0               0
4  2015-01-28   1050614  ...               0.0               0

[5 rows x 48 columns]

for loop로 각 변수에 대하여 미리보기를 할 경우, 변수의 데이터 타입도 동시에 확인할 수 있다. \


*   fecha_dato : 날짜를 기록하는 fecha_dato 변수는 날짜 전용 data type:datetime이 아닌 object이다.
*   age : 고객의 나이를 기록하는 age 변수는 정수값을 가지는 것 같지만, data type이 object이다. 전처리 과정에서 정수값으로의 변환이 필요하다.
*   renta :  가구 총수입을 나타내는 renta 변수는 소수값을 가지면, 5번째 열에서는 결측값을 의미하는 NaN 값을 가진다. 현실에서 사용되는 데이터에서는 이와 같이 다수의 결측값이 존재. 결측값에 대한 변환도 전처리 과정에서 수행한다.

=> 탐색적 데이터 분석을 통해서 얻고자 하는 것은? \
우선 데이터의 일부 행을 랜덤하게 출력하여 눈으로 훑어보며 단계적으로 데이터에 익숙해지고자 한다. 데이터에 대한 간단한 시각화도 도움이 된다.




In [5]:
for col in trn.columns:
    print('{}\n'.format(trn[col].head()))

0    2015-01-28
1    2015-01-28
2    2015-01-28
3    2015-01-28
4    2015-01-28
Name: fecha_dato, dtype: object

0    1375586
1    1050611
2    1050612
3    1050613
4    1050614
Name: ncodpers, dtype: int64

0    N
1    N
2    N
3    N
4    N
Name: ind_empleado, dtype: object

0    ES
1    ES
2    ES
3    ES
4    ES
Name: pais_residencia, dtype: object

0    H
1    V
2    V
3    H
4    V
Name: sexo, dtype: object

0     35
1     23
2     23
3     22
4     23
Name: age, dtype: object

0    2015-01-12
1    2012-08-10
2    2012-08-10
3    2012-08-10
4    2012-08-10
Name: fecha_alta, dtype: object

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: ind_nuevo, dtype: float64

0          6
1         35
2         35
3         35
4         35
Name: antiguedad, dtype: object

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: indrel, dtype: float64

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: ult_fec_cli_1t, dtype: object

0    1
1    1
2    1
3    1
4    1
Name: indrel_1mes, dtype: obje

읽어온 훈련 데이터를 더 자세히 파악하기 위해 pandas에서 자동으로 정보를 요약해 주는 .info()를 사용한다. \
(RangeIndex, Data columns) 값은 행렬 데이터의 열과 행의 크기를 나타낸다. \
훈련 데이터에는 총 13,647,309개의 고객 데이터가 있으며, 고객마다 48개의 변수가 존재한다.

In [6]:
trn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 48 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   fecha_dato             object 
 1   ncodpers               int64  
 2   ind_empleado           object 
 3   pais_residencia        object 
 4   sexo                   object 
 5   age                    object 
 6   fecha_alta             object 
 7   ind_nuevo              float64
 8   antiguedad             object 
 9   indrel                 float64
 10  ult_fec_cli_1t         object 
 11  indrel_1mes            object 
 12  tiprel_1mes            object 
 13  indresi                object 
 14  indext                 object 
 15  conyuemp               object 
 16  canal_entrada          object 
 17  indfall                object 
 18  tipodom                float64
 19  cod_prov               float64
 20  nomprov                object 
 21  ind_actividad_cliente  float64
 22  renta           

**수치형 / 범주형 변수** \
첫 24개의 고객 변수 중, [ "int64", "float64" ]의 데이터 타입을 갖는 수치형 변수를 num_cols로 추출하고, .describe()를 통해 간단한 요약 통계를 확인한다.

In [7]:
num_cols=[col for col in trn.columns[:24] if trn[col].dtype in ['int64','float64']]
trn[num_cols].describe()

ncodpers     ind_nuevo  ...  ind_actividad_cliente         renta
count  1.364731e+07  1.361958e+07  ...           1.361958e+07  1.085293e+07
mean   8.349042e+05  5.956184e-02  ...           4.578105e-01  1.342543e+05
std    4.315650e+05  2.366733e-01  ...           4.982169e-01  2.306202e+05
min    1.588900e+04  0.000000e+00  ...           0.000000e+00  1.202730e+03
25%    4.528130e+05  0.000000e+00  ...           0.000000e+00  6.871098e+04
50%    9.318930e+05  0.000000e+00  ...           0.000000e+00  1.018500e+05
75%    1.199286e+06  0.000000e+00  ...           1.000000e+00  1.559560e+05
max    1.553689e+06  1.000000e+00  ...           1.000000e+00  2.889440e+07

[8 rows x 7 columns]

이번에는 첫 24개의 고객 변수 중, [ "object" ] 데이터 타입을 갖는 범주형 변수를 cat_cols로 추출하고, .describe()를 통해 간단한 요약 통계를 확인한다.

In [8]:
cat_cols = [col for col in trn.columns[:24] if trn[col].dtype in ['O']]
trn[cat_cols].describe()

fecha_dato ind_empleado  ...   nomprov           segmento
count     13647309     13619575  ...  13553718           13457941
unique          17            5  ...        52                  3
top     2016-05-28            N  ...    MADRID  02 - PARTICULARES
freq        931453     13610977  ...   4409600            7960220

[4 rows x 17 columns]

In [9]:
for col in cat_cols:
    uniq = np.unique(trn[col].astype(str))
    print('-'* 50)
    print('# col {}, n_uniq {}, uniq {}'.format(col,len(uniq),uniq))

--------------------------------------------------
# col fecha_dato, n_uniq 17, uniq ['2015-01-28' '2015-02-28' '2015-03-28' '2015-04-28' '2015-05-28'
 '2015-06-28' '2015-07-28' '2015-08-28' '2015-09-28' '2015-10-28'
 '2015-11-28' '2015-12-28' '2016-01-28' '2016-02-28' '2016-03-28'
 '2016-04-28' '2016-05-28']
--------------------------------------------------
# col ind_empleado, n_uniq 6, uniq ['A' 'B' 'F' 'N' 'S' 'nan']
--------------------------------------------------
# col pais_residencia, n_uniq 119, uniq ['AD' 'AE' 'AL' 'AO' 'AR' 'AT' 'AU' 'BA' 'BE' 'BG' 'BM' 'BO' 'BR' 'BY'
 'BZ' 'CA' 'CD' 'CF' 'CG' 'CH' 'CI' 'CL' 'CM' 'CN' 'CO' 'CR' 'CU' 'CZ'
 'DE' 'DJ' 'DK' 'DO' 'DZ' 'EC' 'EE' 'EG' 'ES' 'ET' 'FI' 'FR' 'GA' 'GB'
 'GE' 'GH' 'GI' 'GM' 'GN' 'GQ' 'GR' 'GT' 'GW' 'HK' 'HN' 'HR' 'HU' 'IE'
 'IL' 'IN' 'IS' 'IT' 'JM' 'JP' 'KE' 'KH' 'KR' 'KW' 'KZ' 'LB' 'LT' 'LU'
 'LV' 'LY' 'MA' 'MD' 'MK' 'ML' 'MM' 'MR' 'MT' 'MX' 'MZ' 'NG' 'NI' 'NL'
 'NO' 'NZ' 'OM' 'PA' 'PE' 'PH' 'PK' 'PL' 'PR' 'PT' 'PY' 'Q